### PR TITLE
Stub viewer mode

### PR DESCRIPTION
**Pull request by paulwarren-wk**
#### Reason for change

To reduce the delay between opening a report and the loading mask and progress information being displayed, which can be a significant amount of time on large documents.

This change also reduces the modifications made to the source document, as no `<script>` tags are inserted.  The only  changes made to the source document are the addition of `id` attributes on any `ix` elements that don't already have them.

#### Description of change

The change adds two new command line options:

* `--use-stub-viewer` - if specified, stub viewer mode is enabled.
* `--viewer-suffix` - a suffix that will be inserted into generated filenames.

When using stub viewer mode, the path given to `--save-viewer` should be a directory.  The viewer will output two files:

* `ixbrlviewer.html` - a small HTML file containing the `<script>` reference and JSON data.
* A file with the same name as the input file, with the viewer suffix (if specified) inserted before the file extension.

The viewer  is opened by opening the `ixbrlviewer.html` file.  Due to security restrictions, this needs to be open on a web server (this is the same restriction that applies to Inline XBRL document sets).

#### Steps to Test

Generate a viewer for any file using the command line and adding the `--use-stub-viewer` option and specifying a directory to `--save-viewer`.  Optionally specify a suffix, such as `--viewer-suffix=-viewer`.

This code has been in use on the filings.xbrl.org fork of the viewer for some time, and can be seen on any filing there (e.g. https://filings.xbrl.org/969500N6AVPA51648T62/2022-12-31/ESEF/FR/1/rexel-2022-12-31-fr/reports/ixbrlviewer.html)

**review**:
@Workiva/xt
@paulwarren-wk

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(1) Security relevant changes were detected</summary>
<li>Watched keyword <code>etree</code> in <code>iXBRLViewerPlugin/iXBRLViewer.py</code> line(s) ['27', '339'] added</li>
</details>

## Action Items
* Request infosec review via Slack: #support-infosec
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js
Comment (brettkail-wk) [inline iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js:168]: Out of curiosity, does the hardcoded `.html` force non-XHTML rendering quirks, or does this `_reparentDocument` already cause those to not work (require the input document to be written to the HTML/XHTML subset)?

File: iXBRLViewerPlugin/viewer/src/js/viewer.js
Comment (brettkail-wk) [inline iXBRLViewerPlugin/viewer/src/js/viewer.js:102]: Why is this change required?  It looks like docSetFiles is the same for stub vs default.

File: iXBRLViewerPlugin/__init__.py
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/__init__.py]: > Presumably the reason it's not on by default is due to security. 

Actually, it's just because you can no longer just open the resulting file and have it work.  Adding the note above is a good idea - now done.

> So to be sure, this doesn't reduce the delay for a user to interact with the document, it just causes the "loading" notification to appear more quickly, right?

Pretty much.  I think there is a small win from avoiding the "reparent" step, but it's primarily to get the loading mask up before rather than after the initial document load happens, which can be many seconds on a large document.

File: iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js:168]: The line above is from `master`, not the PR branch.   In stub viewer mode, there is no reparent step.  All documents are loaded via the `src` attribute, and I think rendering mode is determined by the content-type set by the web server, which typically follows the file extension.  So I think that even if the stub viewer is `.html`, the main file gets loaded as XHTML if it has a `.xhtml` extension.

In non-stub mode, I think the reparent step above results in the iframe inheriting rendering mode of the parent.

File: iXBRLViewerPlugin/viewer/src/js/viewer.js
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/viewer/src/js/viewer.js]: In stub mode, docSetFiles is an array containing the iXBRL files, even if there's only one.  In normal mode, docSetFiles is undefined.

I've now moved the length check inside `isDocumentSet()`, and refactored the other code that used this method (30ae2ec) so `isDocumentSet` is back to its original meaning of "is a multi-document document set".

File: iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js
Comment (brettkail-wk) [inline iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js:187]: Pre-existing, I think the `src` needs to somehow URL encode path segments.  Otherwise `Q3: Workiva 10-Q.htm` will be mishandled (`Q3` interpreted as a scheme and normalized to `q3`), or a filename like `javascript:alert(1).html` would become active content.

**Review by brettkail-wk — APPROVED** [review]

File: iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js
Comment (brettkail-wk) [reply iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js:187]: (Approved if you'd rather this be fixed separately.)

Comment (paulwarren-wk) [reply iXBRLViewerPlugin/viewer/src/js/ixbrlviewer.js:187]: Done in [f440403](https://github.com/Workiva/ixbrl-viewer/commit/f440403b388e326c78eea5dddfec3559478bf4a9)

I think it makes sense to do this in viewer generation, rather than the client, and treat this as a list of valid URLs, as this gives us a bit more flexibility in the future if we wanted to be able to put absolute URLs here.

**Comment by soufianeelaissaoui-wk** [comment]
Looks good to me.
Aviary complained about etree, but seems like etree from lxml is safe.
Security +1

File: iXBRLViewerPlugin/__init__.py
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/__init__.py:67]: Should we add these options to the build_viewer.py file?

**Review by brettkail-wk — DISMISSED** [review]

File: iXBRLViewerPlugin/__init__.py
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/__init__.py:67]: Done in [c8a59dd](https://github.com/Workiva/ixbrl-viewer/pull/394/commits/c8a59dd60ca61c8619898a2c7a422c377c83f6e7)

**Review by derekgengenbacher-wf — APPROVED** [review]

**Review by paulyun-wk — APPROVED** [review]
Security +1
for latest commit

**Review by austinmatherne-wk — APPROVED** [review]

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
